File: apps/frontend/src/features/recurring-transactions/commands/recurring-transactions.ts
import { invokeDecodedCommand } from "@/commands/shared";
import type { CommandResult } from "@/commands/shared";
import type { Transaction } from "@/features/transactions/types/model";

import type {
  RecurringBulkAction,
  RecurringBulkExecuteResult,
  RecurringBulkItem,
  RecurringBulkPreflight,
  RecurringMatchingIds,
} from "../types/recurring-bulk";
import type {
  AdoptionPreview,
  BudgetProjectionResult,
  GenerationFailureDiagnostics,
  RecurringAdoptOutcome,
  RecurringCreateOutcome,
  RecurringFailurePage,
  RecurringFeedResult,
  RecurringFeedFilters,
  RecurringFormValues,
  RecurringLifecycleOutcome,
  RecurringMutationOutcome,
  RecurringOccurrencePage,
  RecurringRecoveryOutcome,
  RecurringRepairPreview,
  RecurringRepairField,
  RecurringTransactionDocument,
  ScheduleRule,
  TransactionRecurringProvenance,
} from "../types/recurring-transaction";
import type { RecurringProcessingStatusView } from "../types/recurring-processing-status";
import { RECURRING_COMMANDS } from "./registry";

const toBackendLocal = (value: string): string => {
  if (/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$/.test(value)) {
    return `${value}:00`;
  }
  return value;
};

const toTotalOccurrences = (value: string | undefined): number | null =>
  value ? Number(value) : null;

export const getRecurringBudgetProjections = (input: {
  horizonMonths: number;
  includePausedBudgets?: boolean;
  focusRecurringTransactionId?: string;
}): CommandResult<BudgetProjectionResult> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_budget_projections, {
    horizonMonths: input.horizonMonths,
    includePausedBudgets: input.includePausedBudgets ?? false,
    focusRecurringTransactionId: input.focusRecurringTransactionId,
  });
};

export const buildScheduleRule = (
  values: Pick<
    RecurringFormValues,
    "scheduleKind" | "monthlyDay" | "intervalEvery" | "intervalUnit"
  >,
): ScheduleRule => {
  if (values.scheduleKind === "monthlyDay") {
    return {
      type: "monthlyDay",
      day: Number(values.monthlyDay),
    };
  }
  return {
    type: "interval",
    every: Number(values.intervalEvery),
    unit: values.intervalUnit,
  };
};

export const getRecurringTransactions = (
  limit = 50,
  cursor?: string,
  filters?: RecurringFeedFilters,
): CommandResult<RecurringFeedResult> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_transactions, {
    limit,
    cursor,
    filters,
  });
};

export const getRecurringTransaction = (
  recurringTransactionId: string,
): CommandResult<RecurringTransactionDocument> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_transaction, {
    recurringTransactionId,
  });
};

export const getRecurringTransactionOccurrences = (
  recurringTransactionId: string,
  limit = 50,
  cursor?: string,
): CommandResult<RecurringOccurrencePage> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_transaction_occurrences, {
    recurringTransactionId,
    limit,
    cursor,
  });
};

export const getTransactionRecurringProvenance = (
  transactionId: string,
): CommandResult<TransactionRecurringProvenance | null> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_transaction_recurring_provenance, {
    transactionId,
  });
};

export const createRecurringTransaction = (
  values: RecurringFormValues,
): CommandResult<RecurringCreateOutcome> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.create_recurring_transaction, {
    newRecurringTransaction: {
      schedule: buildScheduleRule(values),
      firstScheduledLocal: toBackendLocal(values.firstScheduledLocal),
      totalOccurrences: toTotalOccurrences(values.totalOccurrences),
      template: {
        description: values.description,
        amount: values.amount,
        transactionType: values.transactionType,
        transactionCategoryId: values.transactionCategoryId || null,
        notes: values.notes || null,
      },
    },
  });
};

export const previewRecurringAdoption = (
  transactionId: string,
  values: Pick<
    RecurringFormValues,
    "scheduleKind" | "monthlyDay" | "intervalEvery" | "intervalUnit" | "totalOccurrences"
  >,
): CommandResult<AdoptionPreview> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.preview_recurring_adoption, {
    request: {
      transactionId,
      schedule: buildScheduleRule(values),
      totalOccurrences: toTotalOccurrences(values.totalOccurrences),
    },
  });
};

export const adoptRecurringTransaction = (
  transaction: Transaction,
  values: RecurringFormValues,
): CommandResult<RecurringAdoptOutcome> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.adopt_recurring_transaction, {
    request: {
      transactionId: transaction.id,
      expectedTransactionDate: toBackendLocal(transaction.transactionDate),
      schedule: buildScheduleRule(values),
      totalOccurrences: toTotalOccurrences(values.totalOccurrences),
      template: {
        description: transaction.description?.trim() || "Recurring transaction",
        amount: transaction.amount,
        transactionType: transaction.transactionType === "income" ? "income" : "expense",
        transactionCategoryId: transaction.transactionCategoryId ?? null,
        notes: transaction.notes?.trim() || null,
      },
    },
  });
};

export const updateRecurringTransaction = (
  document: RecurringTransactionDocument,
  values: RecurringFormValues,
): CommandResult<RecurringMutationOutcome> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.update_recurring_transaction, {
    input: {
      recurringTransactionId: document.recurringTransaction.id,
      expectedRevision: document.recurringTransaction.revision,
      schedule: buildScheduleRule(values),
      nextScheduledLocal: toBackendLocal(values.firstScheduledLocal),
      totalOccurrences: toTotalOccurrences(values.totalOccurrences),
      template: {
        description: values.description,
        amount: values.amount,
        transactionType: values.transactionType,
        transactionCategoryId: values.transactionCategoryId || null,
        notes: values.notes || null,
      },
    },
  });
};

const lifecycleCommand = (
  command:
    | typeof RECURRING_COMMANDS.pause_recurring_transaction
    | typeof RECURRING_COMMANDS.resume_recurring_transaction
    | typeof RECURRING_COMMANDS.stop_recurring_transaction
    | typeof RECURRING_COMMANDS.delete_recurring_transaction,
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringLifecycleOutcome> => {
  return invokeDecodedCommand(command, {
    recurringTransactionId,
    expectedRevision,
  });
};

export const pauseRecurringTransaction = (
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringLifecycleOutcome> => {
  return lifecycleCommand(
    RECURRING_COMMANDS.pause_recurring_transaction,
    recurringTransactionId,
    expectedRevision,
  );
};

export const resumeRecurringTransaction = (
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringLifecycleOutcome> => {
  return lifecycleCommand(
    RECURRING_COMMANDS.resume_recurring_transaction,
    recurringTransactionId,
    expectedRevision,
  );
};

export const stopRecurringTransaction = (
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringLifecycleOutcome> => {
  return lifecycleCommand(
    RECURRING_COMMANDS.stop_recurring_transaction,
    recurringTransactionId,
    expectedRevision,
  );
};

export const deleteRecurringTransaction = (
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringLifecycleOutcome> => {
  return lifecycleCommand(
    RECURRING_COMMANDS.delete_recurring_transaction,
    recurringTransactionId,
    expectedRevision,
  );
};

export const getRecurringTransactionFailureHistory = (
  recurringTransactionId: string,
  limit = 20,
  cursor?: string,
): CommandResult<RecurringFailurePage> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_transaction_failure_history, {
    recurringTransactionId,
    limit,
    cursor,
  });
};

export const getRecurringGenerationFailureDiagnostics = (
  recurringTransactionId: string,
): CommandResult<GenerationFailureDiagnostics> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_generation_failure_diagnostics, {
    recurringTransactionId,
  });
};

export const previewRecurringGenerationRepair = (
  document: RecurringTransactionDocument,
  repairFieldKey: RecurringRepairField,
  values: Pick<
    RecurringFormValues,
    "description" | "amount" | "transactionType" | "transactionCategoryId" | "notes"
  >,
): CommandResult<RecurringRepairPreview> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.preview_recurring_generation_repair, {
    request: {
      recurringTransactionId: document.recurringTransaction.id,
      expectedRevision: document.recurringTransaction.revision,
      repairFieldKey,
      template: {
        description: values.description,
        amount: values.amount,
        transactionType: values.transactionType,
        transactionCategoryId: values.transactionCategoryId || null,
        notes: values.notes || null,
      },
    },
  });
};

export const repairRecurringGenerationFailure = (
  document: RecurringTransactionDocument,
  repairFieldKey: RecurringRepairField,
  values: Pick<
    RecurringFormValues,
    "description" | "amount" | "transactionType" | "transactionCategoryId" | "notes"
  >,
): CommandResult<RecurringRecoveryOutcome> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.repair_recurring_generation_failure, {
    input: {
      recurringTransactionId: document.recurringTransaction.id,
      expectedRevision: document.recurringTransaction.revision,
      repairFieldKey,
      template: {
        description: values.description,
        amount: values.amount,
        transactionType: values.transactionType,
        transactionCategoryId: values.transactionCategoryId || null,
        notes: values.notes || null,
      },
    },
  });
};

export const retryRecurringGenerationFailure = (
  recurringTransactionId: string,
  expectedRevision: number,
): CommandResult<RecurringRecoveryOutcome> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.retry_recurring_generation_failure, {
    input: {
      recurringTransactionId,
      expectedRevision,
    },
  });
};

export const getMatchingRecurringTransactionIds = (
  filters?: RecurringFeedFilters,
): CommandResult<RecurringMatchingIds> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_matching_recurring_transaction_ids, {
    filters,
  });
};

export const preflightRecurringBulk = (
  action: RecurringBulkAction,
  items: Array<RecurringBulkItem>,
): CommandResult<RecurringBulkPreflight> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.preflight_recurring_bulk, {
    request: { action, items },
  });
};

export const executeRecurringBulk = (
  action: RecurringBulkAction,
  items: Array<RecurringBulkItem>,
): CommandResult<RecurringBulkExecuteResult> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.execute_recurring_bulk, {
    request: { action, items },
  });
};

export const getRecurringProcessingStatus = (): CommandResult<RecurringProcessingStatusView> => {
  return invokeDecodedCommand(RECURRING_COMMANDS.get_recurring_processing_status);
};

File: apps/tauri/src/commands/recurring_transactions.rs
use std::sync::Arc;

use log::debug;
use tauri::State;
use zai_app::ServiceContext;
use zai_core::features::recurring_transactions::{
    AdoptRecurringTransaction, AdoptionPreview, AdoptionPreviewRequest,
    GenerationFailureDiagnostics, NewRecurringTransaction, PreviewRecurringGenerationRepair,
    RecurringAdoptOutcome, RecurringBulkExecuteResult, RecurringBulkPreflight,
    RecurringBulkRequest, RecurringCreateOutcome, RecurringFailurePage, RecurringFeedResult,
    RecurringLifecycleOutcome, RecurringLifecycleUpdate, RecurringMatchingIds,
    RecurringMutationOutcome, RecurringOccurrencePage, RecurringRecoveryOutcome,
    RecurringRepairPreview, RecurringTransactionDocument, RepairRecurringGenerationFailure,
    RetryRecurringGenerationFailure, TransactionRecurringProvenance, UpdateRecurringTransaction,
};

use super::{CommandResult, command_error};

#[tauri::command]
pub async fn get_recurring_transactions(
    limit: Option<i64>,
    cursor: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringFeedResult> {
    debug!("Getting recurring transactions feed...");
    state
        .recurring_transactions_service()
        .list_feed(limit, cursor)
        .await
        .map_err(|error| command_error("Failed to load recurring transactions", error))
}

#[tauri::command]
pub async fn get_recurring_transaction(
    recurring_transaction_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringTransactionDocument> {
    debug!(
        "Getting recurring transaction {}...",
        recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .get_document(&recurring_transaction_id)
        .await
        .map_err(|error| command_error("Failed to load recurring transaction", error))
}

#[tauri::command]
pub async fn get_recurring_transaction_occurrences(
    recurring_transaction_id: String,
    limit: Option<i64>,
    cursor: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringOccurrencePage> {
    debug!(
        "Getting occurrences for recurring transaction {}...",
        recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .list_linked_occurrences(&recurring_transaction_id, limit, cursor)
        .await
        .map_err(|error| command_error("Failed to load recurring occurrences", error))
}

#[tauri::command]
pub async fn get_transaction_recurring_provenance(
    transaction_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<Option<TransactionRecurringProvenance>> {
    debug!(
        "Getting recurring provenance for transaction {}...",
        transaction_id
    );
    state
        .recurring_transactions_service()
        .get_transaction_provenance(&transaction_id)
        .await
        .map_err(|error| command_error("Failed to load transaction provenance", error))
}

#[tauri::command]
pub async fn create_recurring_transaction(
    new_recurring_transaction: NewRecurringTransaction,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringCreateOutcome> {
    debug!("Creating recurring transaction...");
    state
        .recurring_transactions_service()
        .create(new_recurring_transaction)
        .await
        .map_err(|error| command_error("Failed to create recurring transaction", error))
}

#[tauri::command]
pub async fn update_recurring_transaction(
    input: UpdateRecurringTransaction,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringMutationOutcome> {
    debug!(
        "Updating recurring transaction {}...",
        input.recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .update(input)
        .await
        .map_err(|error| command_error("Failed to update recurring transaction", error))
}

#[tauri::command]
pub async fn preview_recurring_adoption(
    request: AdoptionPreviewRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<AdoptionPreview> {
    debug!("Previewing recurring adoption...");
    state
        .recurring_transactions_service()
        .preview_adoption(request)
        .await
        .map_err(|error| command_error("Failed to preview adoption", error))
}

#[tauri::command]
pub async fn adopt_recurring_transaction(
    request: AdoptRecurringTransaction,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringAdoptOutcome> {
    debug!("Adopting transaction as recurring occurrence one...");
    state
        .recurring_transactions_service()
        .adopt(request)
        .await
        .map_err(|error| command_error("Failed to adopt transaction", error))
}

#[tauri::command]
pub async fn pause_recurring_transaction(
    recurring_transaction_id: String,
    expected_revision: i32,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringLifecycleOutcome> {
    debug!("Pausing recurring transaction {recurring_transaction_id}...");
    state
        .recurring_transactions_service()
        .pause(RecurringLifecycleUpdate {
            recurring_transaction_id,
            expected_revision,
        })
        .await
        .map_err(|error| command_error("Failed to pause recurring transaction", error))
}

#[tauri::command]
pub async fn resume_recurring_transaction(
    recurring_transaction_id: String,
    expected_revision: i32,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringLifecycleOutcome> {
    debug!("Resuming recurring transaction {recurring_transaction_id}...");
    state
        .recurring_transactions_service()
        .resume(RecurringLifecycleUpdate {
            recurring_transaction_id,
            expected_revision,
        })
        .await
        .map_err(|error| command_error("Failed to resume recurring transaction", error))
}

#[tauri::command]
pub async fn stop_recurring_transaction(
    recurring_transaction_id: String,
    expected_revision: i32,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringLifecycleOutcome> {
    debug!("Stopping recurring transaction {recurring_transaction_id}...");
    state
        .recurring_transactions_service()
        .stop(RecurringLifecycleUpdate {
            recurring_transaction_id,
            expected_revision,
        })
        .await
        .map_err(|error| command_error("Failed to stop recurring transaction", error))
}

#[tauri::command]
pub async fn delete_recurring_transaction(
    recurring_transaction_id: String,
    expected_revision: i32,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringLifecycleOutcome> {
    debug!("Deleting recurring transaction {recurring_transaction_id}...");
    state
        .recurring_transactions_service()
        .delete(RecurringLifecycleUpdate {
            recurring_transaction_id,
            expected_revision,
        })
        .await
        .map_err(|error| command_error("Failed to delete recurring transaction", error))
}

#[tauri::command]
pub async fn preview_recurring_generation_repair(
    request: PreviewRecurringGenerationRepair,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringRepairPreview> {
    debug!(
        "Previewing generation repair for {}...",
        request.recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .preview_generation_repair(request)
        .await
        .map_err(|error| command_error("Failed to preview generation repair", error))
}

#[tauri::command]
pub async fn repair_recurring_generation_failure(
    input: RepairRecurringGenerationFailure,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringRecoveryOutcome> {
    debug!(
        "Repairing generation failure for {}...",
        input.recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .repair_and_retry(input)
        .await
        .map_err(|error| command_error("Failed to repair generation failure", error))
}

#[tauri::command]
pub async fn retry_recurring_generation_failure(
    input: RetryRecurringGenerationFailure,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringRecoveryOutcome> {
    debug!(
        "Retrying generation for {}...",
        input.recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .retry_generation(input)
        .await
        .map_err(|error| command_error("Failed to retry generation", error))
}

#[tauri::command]
pub async fn get_recurring_generation_failure_diagnostics(
    recurring_transaction_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<GenerationFailureDiagnostics> {
    debug!(
        "Copying generation failure diagnostics for {}...",
        recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .generation_failure_diagnostics(&recurring_transaction_id)
        .await
        .map_err(|error| command_error("Failed to load generation failure diagnostics", error))
}

#[tauri::command]
pub async fn get_recurring_transaction_failure_history(
    recurring_transaction_id: String,
    limit: Option<i64>,
    cursor: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringFailurePage> {
    debug!(
        "Loading failure history for recurring transaction {}...",
        recurring_transaction_id
    );
    state
        .recurring_transactions_service()
        .list_failure_history(&recurring_transaction_id, limit, cursor)
        .await
        .map_err(|error| command_error("Failed to load failure history", error))
}

#[tauri::command]
pub async fn get_matching_recurring_transaction_ids(
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringMatchingIds> {
    debug!("Resolving matching recurring transaction ids...");
    state
        .recurring_transactions_service()
        .list_matching_ids()
        .await
        .map_err(|error| command_error("Failed to resolve matching recurring ids", error))
}

#[tauri::command]
pub async fn preflight_recurring_bulk(
    request: RecurringBulkRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringBulkPreflight> {
    debug!("Preflighting recurring bulk action...");
    state
        .recurring_transactions_service()
        .preflight_bulk(request)
        .await
        .map_err(|error| command_error("Failed to preflight recurring bulk action", error))
}

#[tauri::command]
pub async fn execute_recurring_bulk(
    request: RecurringBulkRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> CommandResult<RecurringBulkExecuteResult> {
    debug!("Executing recurring bulk action...");
    state
        .recurring_transactions_service()
        .execute_bulk(request)
        .await
        .map_err(|error| command_error("Failed to execute recurring bulk action", error))
}
